pyqtgraph/functions.py

```diff
@@ -1255,6 +1260,54 @@ def makeARGB(data, lut=None, levels=None, scale=None, useRGBA=False, output=None
     return imgData, alpha
 
 
+def try_fastpath_argb(xp, ain, aout, useRGBA):
+    # we only optimize for certain cases
+    # return False if we did not handle it
+    can_handle = xp is np and ain.dtype == xp.ubyte and ain.flags['C_CONTIGUOUS']
+    if not can_handle:
+        return False
+
+    nrows, ncols = ain.shape[:2]
+    nchans = 1 if ain.ndim == 2 else ain.shape[2]
+
+    # simplify degenerate case
+    if ain.ndim == 3 and nchans == 1:
+        ain = ain.reshape(ain.shape[:2])
+
+    Format = QtGui.QImage.Format
+
+    if nchans == 1:
+        in_fmt = Format.Format_Grayscale8
+    elif nchans == 3:
+        in_fmt = Format.Format_RGB888
+    else:
+        in_fmt = Format.Format_RGBA8888
+
+    if useRGBA:
+        out_fmt = Format.Format_RGBA8888
+    else:
+        out_fmt = Format.Format_ARGB32
+
+    if in_fmt == out_fmt:
+        aout[:] = ain
+        return True
+
+    npixels_chunk = 512*1024
+    batch = int(npixels_chunk / ncols / nchans)
+    batch = max(1, batch)
+    row_beg = 0
+    while row_beg < nrows:
+        row_end = min(row_beg + batch, nrows)
+        ain_view = ain[row_beg:row_end, ...]
+        aout_view = aout[row_beg:row_end, ...]
+        qimg = QtGui.QImage(ain_view, ncols, ain_view.shape[0], ain.strides[0], in_fmt)
+        qimg = qimg.convertToFormat(out_fmt)
```

Comment:
Nothing else in `makeARGB` does batching, though maybe more should. How did you decide on the batch size of 512KPixels? Does this perform well in Windows? ( I'm unable to access my Windows-based test systems today. ) Is this a special case that makes batching particularly valuable, or is it worth exploring this strategy elsewhere?

Reply:
The primary purpose _here_ of the batching was to restrict the amount of extra memory used by going through the fast path due to ```QImage.convertToFormat()``` returning a new image.
So for example, without batching, if the input data was 4096x4096 uint8 (16MB), the ARGB output of ```QImage.convertToFormat()``` would result in the (temporary) consumption of another 64MB. The idea was that the user should not have to pay for more memory usage.

In this case, being cache-friendly was a secondary consideration. However 512KBytes was chosen to be a value that hopefully fit within common processor L2 cache-sizes. It's not too small that the looping overhead becomes dominant. But I would agree with you if it seems like a magic value.

As I said, _here_, it's less for performance than for reducing memory usage. After the speed up here, there were much larger bottlenecks elsewhere that would dwarf any minor gains from further tweaking.

I was trying out this batching strategy within ```rescaleData()``` where it seemed that the multiple operations operating on the same data would benefit, until I found that np.clip() was the issue. I will revisit the batching there some other time. 

In my opinion, the benefit has to be large enough to warrant the addition code complexity / ugliness.

Reply:
Thanks for explaining all that.

Batching might have a pretty substantial performance impact in this case. On a whim, I played with the batch size while processing a 3072x3072 test, and saw:

32-128k: 80fps
256-1024k: 90fps
2048k: 70fps
no batching: 50fps

So at least by that measure, you picked a pretty good magic number! I've only got 256KB L2 cache here, so I don't know how this performs under different conditions.

Reply:
Regarding memory concerns, I just noticed [these docs](https://doc.qt.io/archives/qtforpython-5.14/PySide2/QtGui/QImage.html#PySide2.QtGui.PySide2.QtGui.QImage.convertTo), which says it'll do the conversion in place. It works to replace the last line referenced above with:
```python
        qimg.convertTo(out_fmt)
```

Batching might still be desired here; my system still _reports_ a significantly higher fps when batching at 512k. Yet, I also just now observed that the PySide2 jitter and slowness goes away when batching is disabled. So maybe batching is bad somehow? Ugh, this is still confusing me.